adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java

```diff
@@ -1319,33 +1321,51 @@ public OrderData handleComponentResult(String resultJson) throws Exception {
         PaymentsResponse paymentsResponse = gson.fromJson(resultJson, new TypeToken<PaymentsResponse>() {
         }.getType());
 
-        restoreSessionCart();
-        if (PaymentsResponse.ResultCodeEnum.AUTHORISED == paymentsResponse.getResultCode()) {
-            //TODO: Check for PENDING status.
-            return createAuthorizedOrder(paymentsResponse);
-        }
+        String orderCode = paymentsResponse.getMerchantReference();
+        if(orderCode!=null) {
+            OrderModel orderModel = retrievePendingOrder(orderCode);
+            if (orderModel != null) {
+                if (PaymentsResponse.ResultCodeEnum.AUTHORISED == paymentsResponse.getResultCode()) {
 
-        throw new AdyenNonAuthorizedPaymentException(paymentsResponse);
-    }
 
-    private OrderModel retrievePendingOrder(String orderCode) throws InvalidCartException {
-        OrderModel orderModel = getOrderRepository().getOrderModel(orderCode);
-        if (orderModel == null) {
-            //TODO change exception
-            throw new InvalidCartException("Order '" + orderCode + "' does not exist!");
+                    updateOrderPaymentStatusAndInfo(orderModel, OrderStatus.PAYMENT_AUTHORIZED, paymentsResponse);
+                    OrderData orderData = getOrderConverter().convert(orderModel);
+                    return fillOrderDataWithPaymentInfo(orderData, paymentsResponse);
+                } else {
+                    updateOrderPaymentStatusAndInfo(orderModel, OrderStatus.CANCELLED, paymentsResponse);
+                    restoreCartFromOrder(orderCode);
+                    throw new AdyenNonAuthorizedPaymentException(paymentsResponse);
+                }
+            }
         }
+        throw new AdyenNonAuthorizedPaymentException(paymentsResponse);
+    }
 
-        getSessionService().removeAttribute(SESSION_PENDING_ORDER_CODE);
-        getSessionService().removeAttribute(SESSION_PAYMENT_DATA);
-        getSessionService().removeAttribute(SESSION_MD);
-        getSessionService().removeAttribute(THREEDS2_FINGERPRINT_TOKEN);
-        getSessionService().removeAttribute(THREEDS2_CHALLENGE_TOKEN);
-        getSessionService().removeAttribute(PAYMENT_METHOD);
+    @Override
+    public OrderModel retrievePendingOrder(String orderCode) throws InvalidCartException {
+        OrderModel orderModel;
+        if(orderCode!=null) {
+            orderModel = getOrderRepository().getOrderModel(orderCode);
+            if (orderModel == null) {
+                throw new InvalidCartException("Order '" + orderCode + "' does not exist!");
```

Comment:
I see in the other cases(lines 1361,1414,1419) we do `LOGGER.error` and then `return null`,
while here we throw an exception, is this expected?

Reply:
Its a different scenario and handling is done case by case.